### PR TITLE
Implemented delivery order label

### PR DESCRIPTION
**Pull request by slinkydeveloper**
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

Part of #589

## Proposed Changes

- :gift: Implemented delivery order label
- :bug: Modified default value for delivery order in contract
- :wastebasket: Removed temporarily the usage of upstream features, since they're unstable and causes update issues

**Release Note**

```release-note
Now eventing-kafka-broker implements ordered delivery. You can choose between ordered and unordered delivery through the label kafka.eventing.knative.dev/delivery.order in your triggers. Check out the docs for more details
```

**Docs**

Coming soon in another PR to docs repo

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
I'm going to implement a basic e2e test, although I'm not sure how far I can go with testing this feature in a e2e manner

**Comment by slinkydeveloper** [comment]
Because I would love to use reconciler-test for the e2e, I've opened this one https://github.com/knative-sandbox/eventing-kafka-broker/pull/703

**Comment by pierDipi** [comment]
If you want we can have a status condition to indicate the `DeliveryOrder` configured for a Trigger since it's easy to make a typo and then the `DeliveryOrder` configured is the default one rather than the one users wanted.

**Comment by slinkydeveloper** [comment]
> If you want we can have a status condition to indicate the DeliveryOrder configured for a Trigger since it's easy to make a typo and then the DeliveryOrder configured is the default one rather than the one users wanted. (A separate PR is fine too)

I don't see the point TBH, there's no need to tell the user the controller read the contract properly IMO

**Comment by pierDipi** [comment]
I mean an annotation is weak compared to a key in the spec because we don't have a way to validate that the key is correct for that reason a status condition would be nice to have IMO.

**Comment by slinkydeveloper** [comment]
@pierDipi can we tackle the status issue in another PR? From my understanding (I might be wrong here) we need to setup some alternate condition set, like done in receiver_condition_test.go, right? If that's the case, i think it deserves a separate PR

File: test/e2e_new/trigger/trigger.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/reconciler-test/pkg/feature"
+	"knative.dev/reconciler-test/pkg/k8s"
+	"knative.dev/reconciler-test/pkg/manifest"
+)
+
+// TODO copy pasted from eventing, remove once upstream is clear how templating should work...
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply test/e2e_new/trigger/trigger.go:30]: https://github.com/knative-sandbox/eventing-kafka-broker/issues/731

**Comment by slinkydeveloper** [comment]
/retest

Something wrong in maven?!

**Review by pierDipi — DISMISSED** [review]

**Comment by slinkydeveloper** [comment]
Seems like ordered delivery is not really ordered, will investigate

**Comment by slinkydeveloper** [comment]
The reason why it's failing is veeeery weird. It doesn't add the delivery order in the json:

```
{
  "generation": "38",
  "resources": [
    {
      "uid": "87585205-e53a-442f-8e4b-90192cf2cc0b",
      "topics": [
        "knative-broker-test-oymfewtk-broker-rtbugxnl"
      ],
      "bootstrapServers": "my-cluster-kafka-bootstrap.kafka:9092",
      "ingress": {
        "path": "/test-oymfewtk/broker-rtbugxnl"
      },
      "egresses": [
        {
          "consumerGroup": "5d4fcef4-2551-4f2e-8f7b-88fdf4697015",
          "destination": "http://sink-ngsdgoqr.test-oymfewtk.svc.cluster.local/",
          "uid": "5d4fcef4-2551-4f2e-8f7b-88fdf4697015"
        }
      ]
    }
  ]
}
```

**Comment by slinkydeveloper** [comment]
I've removed broker_test.go, since upstream movement around features is unstable now...

For me we're ready to go
/unhold

**Comment by slinkydeveloper** [comment]
/hold

one sec, maybe it's better to use a label (that's what other people does :smile: )

**Comment by slinkydeveloper** [comment]
/unhold

Ok now we use labels

**Comment by pierDipi** [comment]
I see no benefits in using a label here, basically, the difference is that using a label I can query all objects with that label without doing a full scan or building a custom index whereas an annotation requires a full scan or a custom index.

For example, we use a label in our triggers to get fast access to all triggers associated with a broker. 
https://github.com/knative-sandbox/eventing-kafka-broker/blob/1258e973843974a5a7fc42aff651cab4ed2a3f56/control-plane/pkg/reconciler/trigger/controller.go#L141-L157

In this case, a query like "give me all ordered trigger" seems pretty useless.

Of course, the benefit of using a label comes with additional storage, memory, and compute costs, so my preference would be to use an annotation since this value is only useful in a single reconcile cycle.

**Comment by slinkydeveloper** [comment]
@pierDipi the reason i went for a label is to align with kafka source: https://knative.dev/docs/eventing/samples/kafka/source/#optional-specify-the-key-deserializer

But yeah i was checking in serving and they use annotations too for "special"/"experimental" features, so i'll revert and go for the annotation

**Comment by pierDipi** [comment]
I wonder how this works given the issue on Slack :thinking:

**Review by pierDipi — APPROVED** [review]
Unhold when you're ready!
/hold

**Comment by slinkydeveloper** [comment]
> I wonder how this works given the issue on the partitionkey on Slack thinking

My test tests a single partition :) I'm going to develop a test with different partitions once partitionKey is fixed

**Comment by pierDipi** [comment]
> My test uses a broker with a single partition :) I'm going to develop a test with different partitions once partitionKey is fixed

Oh, gotcha!

**Pull request merged**